app/controllers/tracks/concepts_controller.rb

```diff
@@ -0,0 +1,47 @@
+class Tracks::ConceptsController < ApplicationController
+  before_action :use_track
+
+  allow_unauthenticated! :index, :show
+
+  def authenticated_index
+    use_concepts
+
+    if current_user.joined_track?(@track)
+      render action: "index/joined"
+    else
+      render action: "index/unjoined"
+    end
+  end
+
+  def external_index
+    use_concepts
+  end
+
+  def authenticated_show
+    use_concept
+
+    if current_user.joined_track?(@track)
+      render action: "show/joined"
+    else
+      render action: "show/unjoined"
+    end
+  end
+
+  def external_show
+    use_concept
+  end
+
+  private
+  def use_track
+    @track = Track.find(params[:track_id])
+    @user_track = current_user&.user_track_for(@track)
+  end
+
+  def use_concepts
+    @concepts = @track.concepts
```

Comment:
This will change to generate the Concepts Graph data (cc @neenjaw - this is your entry point for real rendering)

Reply:
for rendering, how is the student's progress communicated from the back end to the front end?

Reply:
We'll expand out the node JSON to include that information. So it'll go either into the graph-data-generator thing you've written, or in a serializer/decorator around it.